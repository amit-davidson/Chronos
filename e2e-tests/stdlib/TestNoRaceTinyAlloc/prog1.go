package main

func main() {
	const P = 4
	const N = 1e6
	var tinySink *byte
	_ = tinySink
	done := make(chan bool)
	for p := 0; p < P; p++ {
		go func() {
			for i := 0; i < N; i++ {
				var b byte
				if b != 0 {
					tinySink = &b // make it heap allocated
				}
				b = 42
			}
			done <- true
		}()
	}
	for p := 0; p < P; p++ {
		<-done
	}
}
