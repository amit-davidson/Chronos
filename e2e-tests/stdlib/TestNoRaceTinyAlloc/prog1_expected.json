{
	"Lockset": {
		"ExistingLocks": {},
		"ExistingUnlocks": {}
	},
	"GuardedAccess": [
		{
			"ID": 1,
			"PosID": 1,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:7:6",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:6:6",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 1,
				"Clock": {
					"1": 1
				}
			}
		},
		{
			"ID": 2,
			"PosID": 2,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:8:2",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:8:14",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 1,
				"Clock": {
					"1": 2
				}
			}
		},
		{
			"ID": 3,
			"PosID": 3,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:8:2",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:8:2",
			"OpKind": 1,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 1,
				"Clock": {
					"1": 3
				}
			}
		},
		{
			"ID": 4,
			"PosID": 4,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:13:8",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:12:9",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 2,
				"Clock": {
					"1": 4,
					"2": 1
				}
			}
		},
		{
			"ID": 5,
			"PosID": 5,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:13:10",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:13:8",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 2,
				"Clock": {
					"1": 4,
					"2": 2
				}
			}
		},
		{
			"ID": 6,
			"PosID": 6,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:13:10",
			"Value": "",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 2,
				"Clock": {
					"1": 4,
					"2": 3
				}
			}
		},
		{
			"ID": 7,
			"PosID": 7,
			"Pos": "",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:13:10",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 2,
				"Clock": {
					"1": 4,
					"2": 4
				}
			}
		},
		{
			"ID": 8,
			"PosID": 8,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:18:4",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:8:2",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 2,
				"Clock": {
					"1": 4,
					"2": 5
				}
			}
		},
		{
			"ID": 9,
			"PosID": 9,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:11:18",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:11:8",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 2,
				"Clock": {
					"1": 4,
					"2": 6
				}
			}
		},
		{
			"ID": 10,
			"PosID": 10,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:11:18",
			"Value": "",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 2,
				"Clock": {
					"1": 4,
					"2": 7
				}
			}
		},
		{
			"ID": 11,
			"PosID": 11,
			"Pos": "",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:11:18",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 2,
				"Clock": {
					"1": 4,
					"2": 8
				}
			}
		},
		{
			"ID": 12,
			"PosID": 12,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:14:6",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:12:9",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 2,
				"Clock": {
					"1": 4,
					"2": 9
				}
			}
		},
		{
			"ID": 13,
			"PosID": 13,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:14:6",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:6:6",
			"OpKind": 1,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 2,
				"Clock": {
					"1": 4,
					"2": 10
				}
			}
		},
		{
			"ID": 14,
			"PosID": 14,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:16:5",
			"Value": "",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 2,
				"Clock": {
					"1": 4,
					"2": 11
				}
			}
		},
		{
			"ID": 15,
			"PosID": 15,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:16:5",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:12:9",
			"OpKind": 1,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 2,
				"Clock": {
					"1": 4,
					"2": 12
				}
			}
		},
		{
			"ID": 16,
			"PosID": 16,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:11:23",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:11:8",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 2,
				"Clock": {
					"1": 4,
					"2": 13
				}
			}
		},
		{
			"ID": 17,
			"PosID": 17,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:11:23",
			"Value": "",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 2,
				"Clock": {
					"1": 4,
					"2": 14
				}
			}
		},
		{
			"ID": 18,
			"PosID": 18,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:9:21",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:9:6",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 1,
				"Clock": {
					"1": 5
				}
			}
		},
		{
			"ID": 19,
			"PosID": 19,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:9:21",
			"Value": "",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 1,
				"Clock": {
					"1": 6
				}
			}
		},
		{
			"ID": 20,
			"PosID": 20,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:9:16",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:9:6",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 1,
				"Clock": {
					"1": 7
				}
			}
		},
		{
			"ID": 21,
			"PosID": 21,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:9:16",
			"Value": "",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 1,
				"Clock": {
					"1": 8
				}
			}
		},
		{
			"ID": 22,
			"PosID": 22,
			"Pos": "",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:9:16",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 1,
				"Clock": {
					"1": 9
				}
			}
		},
		{
			"ID": 23,
			"PosID": 23,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:22:5",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:8:2",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 1,
				"Clock": {
					"1": 10
				}
			}
		},
		{
			"ID": 24,
			"PosID": 24,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:22:3",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:22:5",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 1,
				"Clock": {
					"1": 11
				}
			}
		},
		{
			"ID": 25,
			"PosID": 25,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:21:21",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:21:6",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 1,
				"Clock": {
					"1": 12
				}
			}
		},
		{
			"ID": 26,
			"PosID": 26,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:21:21",
			"Value": "",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 1,
				"Clock": {
					"1": 13
				}
			}
		},
		{
			"ID": 27,
			"PosID": 27,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:21:16",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:21:6",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 1,
				"Clock": {
					"1": 14
				}
			}
		},
		{
			"ID": 28,
			"PosID": 28,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:21:16",
			"Value": "",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 1,
				"Clock": {
					"1": 15
				}
			}
		},
		{
			"ID": 29,
			"PosID": 29,
			"Pos": "",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/stdlib/TestNoRaceTinyAlloc/prog1.go:21:16",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 1,
				"Clock": {
					"1": 16
				}
			}
		}
	]
}